ray-on-gke/tpu/kuberay-tpu-webhook/main.go

```diff
@@ -372,38 +351,79 @@ func getReplicaIndex(clusterName string, groupName string, namespace string) int
 
 // returns next lowest TPU_WORKER_ID in pod slice and updates mappings
 func getNextWorkerID(podSlice slice, namespace string, replicaIndex int) int {
-	tpuWorkerID := 0
-	if sliceToWorkers[podSlice] == nil {
-		newWorker := worker{tpuWorkerID, replicaIndex, true}
-		sliceToWorkers[podSlice] = []worker{newWorker}
-	} else {
-		nextLowestID := math.MaxInt32
-		replacePod := false
-		// iterate through existing workers and check if any have been deleted
-		for _, worker := range sliceToWorkers[podSlice] {
-			if worker.isCreated == false && worker.workerIndex < nextLowestID {
-				replacePod = true
-				nextLowestID = worker.workerIndex
+	tpuWorkerID := 0 // defaults to 0 (first Pod in slice)
+	if sliceToWorkerIDs[podSlice] != nil {
+		sort.Ints(sliceToWorkerIDs[podSlice])
+		// iterate through existing workers and get the next lowest, unused ID
+		for _, workerID := range sliceToWorkerIDs[podSlice] {
+			if workerID == tpuWorkerID {
+				tpuWorkerID++
 			}
 		}
-		// reassign next lowest TPU_WORKER_ID if pod has been deleted
-		if replacePod == true {
-			for index, worker := range sliceToWorkers[podSlice] {
-				// set worker.isCreated to true now that pod is being re-created
-				if worker.workerIndex == nextLowestID {
-					sliceToWorkers[podSlice][index].isCreated = true
+	}
+	klog.V(1).InfoS("getNextWorkerID", "RayCluster", namespace+"/"+podSlice.clusterName, "Worker Group", podSlice.groupName, "TPU_WORKER_ID", tpuWorkerID)
+	return tpuWorkerID
+}
+
+// queries k8s client to build mapping representing the current RayCluster state of TPU pods
+func updateSliceToWorkerIDs(pod *corev1.Pod, clusterName string, groupName string, namespace string, numOfHosts int32) {
+	// retrieve list of Pods in the same namespace and Ray worker group as the intercepted Pod
+	if client == nil {
+		klog.ErrorS(errors.New("k8s go client not initialized"), "updateSliceToWorkerIDs", "RayCluster", namespace+"/"+clusterName)
+		return
+	}
+	listOptions := metav1.ListOptions{
+        LabelSelector: fmt.Sprintf("ray.io/group=%s",groupName),
+    }
+	podsInNamespace, err := client.CoreV1().Pods(namespace).List(context.Background(), listOptions)
+	if err != nil {
+		klog.ErrorS(err, "updateSliceToWorkerIDs", "RayCluster", namespace+"/"+clusterName)
+		return
+	}
+	if podsInNamespace != nil {
+		for _, existingPod := range podsInNamespace.Items {
+			if existingPod.Status.Phase == "Pending" || existingPod.Status.Phase == "Running" {
```

Comment:
nit: do an early `continue` to reduce indentations
```
if existingPod.Status.Phase != "Pending" && existingPod.Status.Phase != "Running" {
    continue
}

existingClusterName := existingPod.Labels["ray.io/cluster"]
existingGroupName := existingPod.Labels["ray.io/group"]
// we only care about workers in the same RayCluster and worker group when assigning IDs
...
```

Reply:
Done in 2192a85 with the change to using a PodInformer.